jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,792 @@
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote address of a {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_ADDRESS = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.getAddress().toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote port of a {@link Request}.
+     * This can be useful if there is an untrusted intermediary, where the remote port can be a surrogate for the connection.
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_PORT = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return Integer.toString(inetSocketAddress.getPort());
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from {@link ConnectionMetaData#getId()} of the {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_CONNECTION = request -> request.getConnectionMetaData().getId();
+
+    private final Map<String, Tracker> _trackers = new ConcurrentHashMap<>();
+    private final Function<Request, String> _getId;
+    private final Tracker.Factory _trackerFactory;
+    private final Request.Handler _rejectHandler;
+    private final int _maxTrackers;
+    private CyclicTimeouts<Tracker> _cyclicTimeouts;
+
+    public DoSHandler()
+    {
+        this(null, 100, -1);
+    }
+
+    /**
+     * @param maxRequestsPerSecond The maximum requests per second allows per ID.
+     */
+    public DoSHandler(@Name("maxRequestsPerSecond") int maxRequestsPerSecond)
+    {
+        this(null, maxRequestsPerSecond, -1);
+    }
+
+    /**
+     * @param trackerFactory Factory to create a Tracker
+     */
+    public DoSHandler(@Name("trackerFactory") Tracker.Factory trackerFactory)
+    {
+        this(null, trackerFactory, null, -1);
+    }
+
+    /**
+     * @param getId Function to extract an remote ID from a request.
+     * @param maxRequestsPerSecond The maximum requests per second allows per ID.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("getId") Function<Request, String> getId,
```

Comment:
`getId` does not mean much, while `remote client identifier function` or `clientIdFn` seems a bit better.

```suggestion
        @Name("clientIdFn") Function<Request, String> clientIdFn,
```